src/Components/Components/src/Rendering/InvalidEventException.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Components.Rendering
+{
+    /// <summary>
+    /// Thrown when the renderer receives an invalid event ID that it can't dispatch.
+    /// </summary>
+    public class InvalidEventException : Exception
```

Comment:
Is the idea of this that the renderer will use more specific types of exceptions to indicate violations of its contract? 

Does this approach scale in practice? Like, are there other places in the renderer where we would define custom exceptions? How many would we end up with?


Should this inherit from `ArgumentException`?

Reply:
Sort of.

* We need to differentiate bad input from errors in the app, as we want to log them with different importance (and in the future we might decide to do different things).
* We need an exception we can throw in Components when we parse the event arguments or when we receive an event for which there's no associated event handler (both of which can be caused by bad input).

I wouldn't extend ArgumentException, this exception is thrown in response to an argument not being valid, but we are the only ones to throw and consume this exception and having the argument information doesn't add any value.

To go even further, I would even put this exception in the .Internal namespace if we can, as its purely infrastructure.

WRT scalability of the app, there are only three other places where we interact with user code.
* When we instantiate a component.
* When we run the BuildRenderTree (or equivalent) function from the component.
* When we run the handler for an event.


We currently don't explicitly catch exceptions in those call-sites (except for the handler). We simply let the exception bubble up (ultimately reaching ProcessRenderQueue) and we handle it there.

I don't think there's any special value in throwing a custom exception in those cases as we always do the same thing for exceptions in those cases.
* Log a warning.
* Terminate the circuit.

Accepting that the renderer is not something we expect customers to extend/replace, we could if we wanted introduce a new exception in the future, when we find a reason it adds value.

At that point, I would simply catch all exceptions at the call-sites and would wrap them inside a ComponentException or similar, but even if we were to create an exception hierarchy we woud have at most two or 3 new types to represent all the error space, I think:
* Exception in constructor.
* Exception injecting dependencies. (Could be folded into the one above)
* Exception rendering the component.
* Exception setting the parameters for the component.

That said, I don't think we need to be specific about any of the above as they all result in the same action.

Reply:
I like the overall design here.

Regarding making this `public`, that's just because it has to be referenced in `.Server`, right? But the `.Server` project doesn't do anything more specific with `InvalidEventException` other than recognize its type?

If this is the case, instead of each of the exception types being `public`, could we have one `public` base class for them (e.g., `InvalidIncomingMessageException`)? Then that's the only thing `.Server` has to look for, and we can have as many `internal` subclasses as we want without it being visible on the public API surface.

Reply:
Also, on naming, could we do something to be super clear that this means *incoming data was invalid*, as opposed to some other sense in which the event was "invalid"? For example, `InvalidIncomingEventDataException`.

Reply:
@SteveSandersonMS Here are some thoughts:
* I wanted to avoid having something that was very specific to the server, like someone sent you bad input.
* The InvalidEvent sounded general enough to be thrown in the cases where we care about.
  * I don't expect us to have many of these.
  * I believe introducing a base class is not a breaking change, so we can do it later if we need to.
  * I see your point about IncommingData vs InvalidEventException, but I would like to see if we can find a better name. Maybe InvalidEventDescriptorException?

Reply:
Would a marker interface play the same role here?

Reply:
Got rid of this, as we don't need it with Ryan's changes